src/mgxs_data.F90

```diff
@@ -0,0 +1,705 @@
+      if (check_for_node(node_xsdata, "multiplicity")) then
+        arr_len = get_arraysize_double(node_xsdata, "multiplicity")
+        if (arr_len == groups * groups * this % Nazi * this % Npol) then
+          allocate(temp_arr(arr_len))
+          call get_node_array(node_xsdata, "multiplicity", temp_arr)
+          this % mult = reshape(temp_arr, (/groups, groups, this % Nazi, this % Npol/))
+          deallocate(temp_arr)
+        else
+          error_code = 1
+          error_text = "Multiplicity Length Does Not Match!"
+          return
+        end if
+      else
+        this % mult = ONE
+      end if
+
+    end subroutine NuclideAngle_init
+
+
+!===============================================================================
+! CREATE_MACRO_XS generates the macroscopic x/s from the microscopic input data
+!===============================================================================
+
+  subroutine create_macro_xs()
+    integer :: i_mat ! index in materials array
+    integer :: i             ! loop index over nuclides
+    integer :: l             ! Loop over score bins
+    type(Material), pointer :: mat ! current material
+    logical :: get_kfiss, get_fiss
+    integer :: error_code
+    character(MAX_LINE_LEN) :: error_text
+    integer :: representation
+    integer :: scatt_type
+    integer :: legendre_mu_points
+
+    ! Find out if we need fission & kappa fission
+    ! (i.e., are there any SCORE_FISSION or SCORE_KAPPA_FISSION tallies?)
+    get_kfiss = .false.
+    get_fiss  = .false.
+    do i = 1, n_tallies
+      do l = 1, tallies(i) % n_score_bins
+        if (tallies(i) % score_bins(l) == SCORE_KAPPA_FISSION) then
+          get_kfiss = .true.
+        end if
+        if (tallies(i) % score_bins(l) == SCORE_FISSION) then
+          get_fiss = .true.
+        end if
+      end do
+      if (get_kfiss .and. get_fiss) &
+           exit
+    end do
+
+    allocate(macro_xs(n_materials))
+
+    do i_mat = 1, n_materials
+      mat => materials(i_mat)
+
+      ! Check to see how our nuclides are represented
+      ! Force all to be the same type
+      ! Therefore type(nuclides(mat % nuclide(1)) % obj) dictates type(macroxs)
+      ! At the same time, we will find the scattering type, as that will dictate
+      ! how we allocate the scatter object within macroxs
+      legendre_mu_points = nuclides_MG(mat % nuclide(1)) % obj % legendre_mu_points
+      select type(nuc => nuclides_MG(mat % nuclide(1)) % obj)
+      type is (NuclideIso)
+        representation = MGXS_ISOTROPIC
+      type is (NuclideAngle)
+        representation = MGXS_ANGLE
+      end select
+      scatt_type = nuclides_MG(mat % nuclide(1)) % obj % scatt_type
+
+      ! Now allocate accordingly
+      select case(representation)
+      case(MGXS_ISOTROPIC)
+        allocate(MacroXSIso :: macro_xs(i_mat) % obj)
+      case(MGXS_ANGLE)
+        allocate(MacroXSAngle :: macro_xs(i_mat) % obj)
```

Comment:
Couldn't you just do these allocations in the `select type` above rather than setting representation and checking it?